### PR TITLE
doc/inference api

### PR DESCRIPTION
**Pull request by Superjomn**

### CODE REVIEW

File: paddle/contrib/inference/high_level_api.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+The definition is 
+
+```c++
+enum PaddleDType {
+  FLOAT32,
+  INT64,
+};
+
+struct PaddleBuf {
+  void* data;     // pointer to the data memory.
```
Comment (panyx0718) [inline paddle/contrib/inference/high_level_api.md]: need to document the data layout. Is it row-major?

File: paddle/contrib/inference/high_level_api.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  // kAutoMixedAnakin,    // Automatically mix Fluid with Anakin.
+  // kAutoMixedTensorRT,  // Automatically mix Fluid with TensorRT.
+};
+```
+
+## PaddlePredictor and how to create one
+The main interface is `PaddlePredictor`, there are following methods 
+
+- `bool Run(const std::vector<PaddleTensor>& inputs, std::vector<PaddleTensor>* output_data)`
+  - take inputs and output `output_data`
```
Comment (panyx0718) [inline paddle/contrib/inference/high_level_api.md:44]: Document pointer ownership?

Comment (panyx0718) [reply paddle/contrib/inference/high_level_api.md:44]: Document thread-safety?

File: paddle/contrib/inference/high_level_api.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
+# Inference High-level APIs
+This document describes the high-level inference APIs one can use to easily deploy a Paddle model for an application.
+
+The APIs are described in `paddle_inference_api.h`, just one header file, and two libaries `libpaddle_fluid.so` and `libpaddle_fluid_api.so` are needed.
```
Comment (luotao1) [inline paddle/contrib/inference/high_level_api.md:4]: 对api来说，提供一个`libpaddle_fluid_api.so`是否就可以了？

File: paddle/contrib/inference/high_level_api.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+## PaddleTensor
+Forget the hard-to-understand `LoDTensor`, 
+we provide the `PaddleTensor` data structure is to give a general tensor interface.
```
Comment (luotao1) [inline paddle/contrib/inference/high_level_api.md]: 8行有两个谓语 provide和 is

File: paddle/contrib/inference/high_level_api.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  std::string name;  // variable name.
+  std::vector<int> shape;
+  PaddleBuf data;  // blob of data.
+  PaddleDType dtype;
+};
+```
+
+The data is stored in a continuous memory `PaddleBuf`, and tensor's data type is specified by a `PaddleDType`. 
+The `name` field is used to specify the name of input variable, 
+that is important when there are multiple inputs and need to distiuish which variable to set the content.
```
Comment (luotao1) [inline paddle/contrib/inference/high_level_api.md]: which variable to set the content 缺少谓语

File: paddle/contrib/inference/high_level_api.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+## engine
+The inference APIs has different underlying implementation, currently there are two valid engines:
```
Comment (luotao1) [inline paddle/contrib/inference/high_level_api.md]: have different underlying implementations

File: paddle/contrib/inference/high_level_api.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+- the native engine, which is consists of the native operators and framework,
```
Comment (luotao1) [inline paddle/contrib/inference/high_level_api.md]: is consists of 语法请修正

File: paddle/contrib/inference/high_level_api.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+- the Anakin engine, which is a Anakin library embeded.
```
Comment (luotao1) [inline paddle/contrib/inference/high_level_api.md]: 这里可以给个Anakin的链接

File: paddle/contrib/inference/high_level_api.md [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+The native engine takes a native Paddle model as input, and supports any model that trained by Paddle.
+
+The Anakin engine can only take the Anakin model as input(user need to manully transform the format first) and currently not all Paddle models are supported.
```
Comment (luotao1) [inline paddle/contrib/inference/high_level_api.md]: 41行需要放到38行后面，43行需要放到39行后面么？

File: paddle/contrib/inference/high_level_api.md [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+```
+
+## PaddlePredictor and how to create one
+The main interface is `PaddlePredictor`, there are following methods 
+
+- `bool Run(const std::vector<PaddleTensor>& inputs, std::vector<PaddleTensor>* output_data)`
+  - take inputs and output `output_data`
+- `Clone` to clone a predictor from an existing one, with model parameter shared.
+
+There is a factory method to help create a predictor
```
Comment (luotao1) [inline paddle/contrib/inference/high_level_api.md]: 缺句号。

File: paddle/contrib/inference/high_level_api.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+template <typename ConfigT, PaddleEngineKind engine = PaddleEngineKind::kNative>
+std::unique_ptr<PaddlePredictor> CreatePaddlePredictor(const ConfigT& config);
+```
+
+By specify the engine kind and config, one can get an specific implementation.
+
+## Reference
+
+- [paddle_inference_api.h](./paddle_inference_api.h)
+- [demos](./demo)
```
Comment (luotao1) [inline paddle/contrib/inference/high_level_api.md:59]: demo里还需要加文档，类似https://github.com/PaddlePaddle/Paddle/tree/develop/paddle/fluid/train/demo

**Review by luotao1 — COMMENTED** [review]
high_level_api.md 需要挪到doc/fluid下面么？

File: paddle/contrib/inference/high_level_api.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
+# Inference High-level APIs
+This document describes the high-level inference APIs one can use to easily deploy a Paddle model for an application.
+
+The APIs are described in `paddle_inference_api.h`, just one header file, and two libaries `libpaddle_fluid.so` and `libpaddle_fluid_api.so` are needed.
+
+## PaddleTensor
+Forget the hard-to-understand `LoDTensor`,
```
Comment (panyx0718) [inline paddle/contrib/inference/high_level_api.md]: Can remove this line

File: paddle/contrib/inference/high_level_api.md (hunk 3 above)
Comment (Superjomn) [reply paddle/contrib/inference/high_level_api.md:4]: 可以编译进去吗，我之前发现不行，还需要 paddle_fluid.so

File: paddle/contrib/inference/high_level_api.md (hunk 7 above)
Comment (Superjomn) [reply paddle/contrib/inference/high_level_api.md]: 这句没问题啊

File: paddle/contrib/inference/high_level_api.md (hunk 11 above)
Comment (Superjomn) [reply paddle/contrib/inference/high_level_api.md:59]: 在另外一个 pr 里加

File: paddle/contrib/inference/high_level_api.md (hunk 9 above)
Comment (Superjomn) [reply paddle/contrib/inference/high_level_api.md]: done

File: paddle/contrib/inference/high_level_api.md (hunk 2 above)
Comment (Superjomn) [reply paddle/contrib/inference/high_level_api.md:44]: The return is a `unique_ptr`,  the ownership will transparently transfer.

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**